### PR TITLE
Assignment6

### PR DESCRIPTION
**Pull request by Bhopss152**
Joshua Hess
j0shh3ss

### CODE REVIEW

**Comment by jstanley6** [comment]
I will review

**Review by jstanley6 — COMMENTED** [review]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ❌
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ❌

Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ❌
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
and enabled .NET analyzers for both projects ❌
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

File: Calculate/Program.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿using Calculate;
+namespace Calculate;
+using System;
+
+
+class Program
+{
+    static void Main()
```
Comment (jstanley6) [inline Calculate/Program.cs:8]: There is no empty constructor that sets your WriteLine ReadLine init only properties. Like
    Writeline = Console.WriteLine;
        Readline = Console.ReadLine;

File: Calculate/Program.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        var calculator = new Calculator();
+
+        while (true)
+        {
+            Console.WriteLine("Enter function in form x + y");
+            string input = Console.ReadLine();
+            if (string.IsNullOrWhiteSpace(input)) break;
+
+            if (calculator.TryCalculate(input, out int result))
```
Comment (jstanley6) [inline Calculate/Program.cs:18]: You aren't using the ReadLine Properties that you have set in the program class. You are only using the System ReadLine WriteLine library

File: CalculateTests/CalculateTests.csproj [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.3.2" />
+    <PackageReference Include="MSTest.TestAdapter" Version="2.2.10" />
+    <PackageReference Include="MSTest.TestFramework" Version="2.2.10" />
+    <PackageReference Include="coverlet.collector" Version="3.1.2" />
+  </ItemGroup>
+
+	<ItemGroup>
+		<ProjectReference Include="..\Calculate\Calculate.csproj" />
```
Comment (jstanley6) [inline CalculateTests/CalculateTests.csproj:19]: All this should be in a Directory.Build.props file and the Langversion needs set as well as a few other things that were missed in the instructions. Check the Fundamentals section and it should give you more insight on what you missed setting. There is a few warnings that I was getting which should help you to adjust the console application.

File: CalculateTests/CalculatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            // Arrange
+            Calculator calculator = new Calculator();
+            string input = "10 / 2";
+            int expected = 5;
+
+            // Act
+            bool success = calculator.TryCalculate(input, out int result);
+
+            // Assert
```
Comment (jstanley6) [inline CalculateTests/CalculatorTests.cs:68]: Can use more tests such as the first test to initialize the WriteLine and ReadLine properties at construction time. 
  [TestMethod]
    public void TestWriteLine()
    {

File: Calculate/Program.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            else
+            {
+                Console.WriteLine("Invalid input");
+            }
+        }
+    }
+ 
+    public  Action<string> WriteLine { get; init; } = Console.WriteLine;
+    public  Func<string> ReadLine { get; init; } = Console.ReadLine;
+}
```
Comment (Joshua-Pedersen) [inline Calculate/Program.cs:31]: You don't seem to have any tests for your ``Program.cs`` file. Could test to assure that your ``ReadLine`` and ``WriteLine`` and working correctly.

File: CalculateTests/CalculateTests.csproj [hunk 6]
```diff
@@ -0,0 +1,22 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net7.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+
```
Comment (Joshua-Pedersen) [inline CalculateTests/CalculateTests.csproj:7]: Need to add ``<LangVersion>`` also need to create a ``Directory.Build.Props`` file where you can put all the PropertyGroup info.

File: Calculate/Calculator.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            ['+'] = Add,
+            ['-'] = Subtract,
+            ['*'] = Multiply,
+            ['/'] = Divide
+        };
+
+        private static int Add(int x, int y) => x + y;
+        private static int Subtract(int x, int y) => x - y;
+        private static int Multiply(int x, int y) => x * y;
+        private static int Divide(int x, int y) => y != 0 ? x / y : throw new DivideByZeroException();
```
Comment (Joshua-Pedersen) [inline Calculate/Calculator.cs:22]: I like the addition of a divide by zero catch

**Review by Joshua-Pedersen — COMMENTED** [review]
## Instructions

- Create a *Console* project called "Calculate.". ✔
- Define a Program Class
  - Define two init-only setter properties, `WriteLine` and `ReadLine`, that contain delegates for writing a line of text and reading a line of text respectively ❌
  - Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌
  - Set the default behavior for the `WriteLine` and `ReadLine` properties to invoke `System.Console` versions of the methods and add an empty default constructor. ✔
- Define a Calculator class ✔
  - Define static `Add`, `Subtract`, `Multiple`, and `Divide` methods that have two parameters and return a third parameter. ✔
  - Define a read-only property, `MathematicalOperations`, of type `System.Collections.Generics.IReadOnlyDictionary<TKey,TValue>` that:
    - is initialized to a `System.Collections.Generics.Dictionary<<TKey,TValue>` instance that. ✔
      - Uses `char` for the key corresponding to the operators +, -, *, and /. ✔
      - Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
  - Implement a `TryCalculate` method following "TryParse" pattern ✔
    - Valid `calculation` expressions include such strings as "3 + 4", "42 - 2", etc. ✔
    - If there is no whitespace around the operator, you can assume the `calculation` is invalid and return false. Similarly if the operands are not integers. ✔
    - Use `string.Split()`, pattern matching, logical and operators to parse the string in their entirety ✔
    - Index into the `MathematicalOperations` method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
- Be sure to use the `WriteLine`/`ReadLine` properties on `Program` for testing the input and output of your program. ❌

## Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

- Refactor the redirect portion of the `Program` class into 'ProgramBase`
- Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
- Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file. ❌
- Place all shared project items into a `Directory.Build.targets` file. ❌
- nullable reference types is enabled  ❌
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ❌
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)   ❌
- and enabled .NET analyzers for both projects ❌
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ❌
- Choose simplicity over complexity ✔

File: Calculate/Program.cs (hunk 1 above)
Comment (j0shh3ss) [reply Calculate/Program.cs:8]: Yes there is via line 29,30

File: Calculate/Program.cs (hunk 2 above)
Comment (j0shh3ss) [reply Calculate/Program.cs:18]: Yes we are, via line 29,30

File: Calculate/Program.cs (hunk 1 above)
Comment (Bhopss152) [reply Calculate/Program.cs:8]: added and fixed thank you resolving conversation

File: CalculateTests/CalculatorTests.cs (hunk 4 above)
Comment (Bhopss152) [reply CalculateTests/CalculatorTests.cs:68]: Tried to resolve.

File: Calculate/Program.cs (hunk 5 above)
Comment (Bhopss152) [reply Calculate/Program.cs:31]: Tried to resolve

File: Calculate/Calculator.cs (hunk 7 above)
Comment (Bhopss152) [reply Calculate/Calculator.cs:22]: Thanks!

File: Calculate/Program.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        program.Run(calculator);
+    }
+
+    public Action<string> WriteLine { get; set; }
+    public Func<string> ReadLine { get; set; }
+    public void SetUpInput(Action<string> writeLine, Func<string> readLine)
+    {
+        WriteLine = writeLine;
+        ReadLine = readLine;
+    }
```
Comment (mmwoodfo) [inline Calculate/Program.cs:19]:
```suggestion
    public Program(Action<string> writeLine, Func<string> readLine)
    {
        WriteLine = writeLine;
        ReadLine = readLine;
    }

    public Program()
    {
        WriteLine = Console.WriteLine;
        ReadLine = Console.ReadLine;
    }
```
You should really be doing something line this

File: Calculate/Calculator.cs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    public bool TryCalculate(string input, out int result)
+    {
+        result = 0;
+
+        if (string.IsNullOrWhiteSpace(input))
+        {
+            return false;
+        }
+
+        string[] parts = input.Split(' ');
```
Comment (mmwoodfo) [inline Calculate/Calculator.cs:25]: You should be using the Split method along side pattern matching. e.g.
`if (expression?.Split(" ") is [string operator1, [char mathmaticalOperator], string operand2])`

File: Calculate/Calculator.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            { "+", (x, y) => x + y },
+            { "-", (x, y) => x - y },
+            { "*", (x, y) => x * y },
+            { "/", (x, y) => x / y }
+        };
+    }
+
+    public bool TryCalculate(string input, out int result)
+    {
+        result = 0;
```
Comment (mmwoodfo) [inline Calculate/Calculator.cs:18]: Consider making this nullable, and returning null here if the calculation is not successful.

File: CalculateTests/ProgramTests.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
+// We really tried to implement testers for writeline and readline but could not get to function correctly.
+//using Microsoft.VisualStudio.TestPlatform.TestHost;
+//using Microsoft.VisualStudio.TestTools.UnitTesting;
+//using System;
+
+//namespace CalculatorTests
```
Comment (mmwoodfo) [inline CalculateTests/ProgramTests.cs:6]: Were these failing?

**Pull request closed without merging**